src/EachPromise.php

```diff
@@ -169,11 +173,21 @@ function ($reason) use ($idx) {
 
     private function advanceIterator()
     {
+        // Place a lock on the iterator so that we ensure to not recurse,
+        // preventing fatal generator errors.
+        if ($this->mutex) {
+            return false;
+        }
+
+        $this->mutex = true;
+
         try {
             $this->iterable->next();
+            $this->mutex = false;
```

Comment:
Nitpick, but I think your intention would be clearer if this line were in a `finally` block instead of both the `try` and `catch`.